src/platform/x86_64/vmx/vmx.c

```diff
@@ -0,0 +1,412 @@
+#include <chal_cpu.h>
+#include <thd.h>
+#include <cpuid.h>
+#include <chal_asm_inc.h>
+#include <vm.h>
+#include <vmx.h>
+#include <vmx_msr.h>
+#include <vmx_logging.h>
+#include <vmx_exit.h>
+#include <vmx_utils.h>
+#include <vmx_vmcs.h>
+
+extern u64_t get_idt_base(void);
+extern u64_t get_tss_base(cpuid_t cpu_id);
+
+static inline u64_t get_fs_base(void) { return msr_get(IA32_FS_BASE); }
+static inline u64_t get_gs_base(void){ return msr_get(IA32_GS_BASE); }
+static inline u64_t get_efer(void) { return msr_get(IA32_EFER); }
+static inline struct vmx_vmcs *get_vmcs(struct vm_vcpu_context *vcpu_ctx) { return &vcpu_ctx->vmcs; }
+
+u64_t cr4_fixed1_bits = 0;
+u64_t cr4_fixed0_bits = 0;
+u64_t cr0_fixed1_bits = 0;
+u64_t cr0_fixed0_bits = 0;
+
+/* Note: this page is just used for enabling cpu vm capability, it doesn't relate to vcpu VMCS pages */
+static char vm_env_page[PAGE_SIZE_4K] __attribute__((aligned(PAGE_SIZE_4K)));
+
+void
+vmx_env_init(void)
+{
+	assert(sizeof(struct vm_vcpu_shared_region) < PAGE_SIZE_4K);
+	memset(&vm_env_page, 0, PAGE_SIZE_4K);
+	vmx_on(&vm_env_page);
+}
+
+void
+vmx_host_state_init(void)
+{
+	u64_t tmp64;
+	u32_t eax, edx;
+
+	CR_GET(cr0, &tmp64);
+	vmwrite(HOST_CR0, tmp64);
+	VMX_DEBUG("VMX_HOST_CR0: 0x%p\n", tmp64);
+
+	CR_GET(cr3, &tmp64);
+	vmwrite(HOST_CR3, tmp64);
+	VMX_DEBUG("VMX_HOST_CR3: 0x%p\n", tmp64);
+
+	CR_GET(cr4, &tmp64);
+	vmwrite(HOST_CR4, tmp64);
+	VMX_DEBUG("VMX_HOST_CR4: 0x%p\n", tmp64);
+
+	/* The sp and ip used when VM-exit */
+	vmwrite(HOST_RSP, (u64_t)kernel_stack_info[get_cpuid()].kernel_stack_addr);
+	vmwrite(HOST_RIP, (u64_t)&vmx_exit_handler_asm);
+	
+	vmwrite(HOST_DS, SEL_KDSEG);
+	vmwrite(HOST_ES, SEL_KDSEG);
+	vmwrite(HOST_FS, SEL_KDSEG);
+	vmwrite(HOST_GS, SEL_KDSEG);
+	vmwrite(HOST_TR, SEL_TSS);
+	vmwrite(HOST_FS_BASE, get_fs_base());
```

Comment:
The fs base is part of the thread switch process. Yes, thus it will change. However, since a vcpu is implemented as a thread, thus when switching vcpu/threads, the fs will be save and restored automatically. I cannot think of any issues here for now.